File: src/main/java/com/sarthak/agenticai/service/ProductService.java
    package com.sarthak.agenticai.service;

    import com.sarthak.agenticai.dto.ProductRequestDto;
    import com.sarthak.agenticai.dto.ProductResponseDto;
    import org.springframework.data.domain.Page;
    import com.sarthak.agenticai.projection.ProductSummary;
    import com.sarthak.agenticai.dto.ProductSummaryDto;

    import java.util.List;

    public interface ProductService {

        ProductResponseDto createProduct(ProductRequestDto request);
        Page<ProductResponseDto> getAllProducts(
                int pageNumber,
                int pageSize,
                String sortBy,
                String sortDir
        );

        ProductResponseDto getProductById(Long id);
        List<ProductResponseDto> searchProducts(String keyword);
        List<ProductResponseDto> getProductsByCategory(Long categoryId);
        List<ProductResponseDto> filterProductsByPrice(
                Double minPrice,
                Double maxPrice);

        ProductResponseDto updateProduct(Long id,
                                         ProductRequestDto request);

        void deleteProduct(Long id);
        Page<ProductResponseDto> filterProducts(

                String keyword,

                Long categoryId,

                Double minPrice,

                Double maxPrice,

                int pageNumber,

                int pageSize,

                String sortBy,

                String sortDir
        );
        List<ProductSummary> getProductSummary();
        List<ProductSummaryDto> getProductSummaryDto();
        List<ProductResponseDto> getRecommendedProducts();

        List<ProductResponseDto> getTopCheapProducts();

        List<ProductResponseDto> getTopExpensiveProducts();

        List<ProductResponseDto> getAvailableProducts();

        List<ProductResponseDto> getProductsByCategoryName(String categoryName);

    }

File: agentic-ai-frontend/src/services/productService.js
import api from "../api/axios";

// ========================================
// GET ALL PRODUCTS
// ========================================

export const getAllProducts = async (
    pageNumber = 0,
    pageSize = 5,
    sortBy = "id",
    sortDir = "asc"
) => {

    const response = await api.get("/products", {
        params: {
            pageNumber,
            pageSize,
            sortBy,
            sortDir
        }
    });

    return response.data;
};


// ========================================
// GET PRODUCT BY ID
// ========================================

export const getProductById = async (id) => {

    const response = await api.get(`/products/${id}`);

    return response.data;
};


// ========================================
// SEARCH PRODUCTS
// ========================================

export const searchProducts = async (keyword) => {

    const response = await api.get("/products/search", {
        params: {
            keyword
        }
    });

    return response.data;
};


// ========================================
// GET PRODUCTS BY CATEGORY
// ========================================

export const getProductsByCategory = async (categoryId) => {

    const response = await api.get(
        `/products/category/${categoryId}`
    );

    return response.data;
};


// ========================================
// FILTER PRODUCTS BY PRICE
// ========================================

export const filterProductsByPrice = async (
    minPrice,
    maxPrice
) => {

    const response = await api.get("/products/filter", {
        params: {
            minPrice,
            maxPrice
        }
    });

    return response.data;
};


// ========================================
// GET ALL CATEGORIES
// ========================================

export const getAllCategories = async () => {

    const response = await api.get("/categories");

    return response.data;
};